File: 백준/Gold/15686. 치킨 배달/치킨 배달.py
N, M = map(int, input().split())
arr = [list(map(int, input().split())) for _ in range(N)]
chicken = []
house = []
lst = []
visited = []
result = 50*50

for i in range(N):
    for j in range(N):
        if arr[i][j] == 1:
            house.append([i, j])

        if arr[i][j] == 2:
            chicken.append([i, j])
            visited.append([False])

def sol(n):
    global result
    if len(lst) == M:
        # print(lst)
        tmp_sum = 0
        for h in house:
            tmp = 50 * 50
            for c in lst:
            # for h in house:
                distance = abs(h[0] - c[0]) + abs(h[1] - c[1])
                if distance < tmp:
                    tmp = distance
            tmp_sum += tmp
        # print(tmp_sum)
        if tmp_sum < result:
            result = tmp_sum
        return

    else:
        for i in range(n, len(chicken)):
            # if not visited[i]:
            lst.append(chicken[i])
            # visited[i] = True
            sol(i+1)
            lst.pop()
            # visited[i] = False

sol(0)
print(result)

File: 백준/Gold/15686. 치킨 배달/치킨 배달.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.StringTokenizer;

public class Main {
    static int N, M;
    static int result = Integer.MAX_VALUE;
    static ArrayList<int[]> house = new ArrayList<>();;
    static ArrayList<int[]> chicken = new ArrayList<>();;
    static boolean[] selected;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;
        st = new StringTokenizer(br.readLine());
        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());

        // 집, 치킨집의 좌표 담은 배열
        for (int i = 0; i < N; i++) {
            st = new StringTokenizer(br.readLine());
            for (int j = 0; j < N; j++) {
                int num = Integer.parseInt(st.nextToken());
                if(num == 1) house.add(new int[]{i, j});
                else if(num == 2) chicken.add(new int[]{i, j});
            }
        }
        // 선택 여부 확인 위한 치킨집 개수만큼의 배열 selected 생성
        selected = new boolean[chicken.size()];

        sol(0, 0);
        System.out.println(result);
    }

    static void sol(int cnt, int idx) {
        // M개만큼 선택되었다면 각 집별로 선택된 치킨집에 대한 치킨 거리 계산 및 최소값 판단
        if (cnt == M) {
            int sum = 0;
            for (int[] h : house) {
                int dist = Integer.MAX_VALUE;
                for (int j = 0; j < chicken.size(); j++) {
                    if (selected[j]) {
                        int tmp = Math.abs(chicken.get(j)[0] - h[0]) + Math.abs(chicken.get(j)[1] - h[1]);
                        dist = Math.min(dist, tmp);
                    }
                }
                sum += dist;
            }
            result = Math.min(result, sum);
            return;
        }

        for (int i = idx; i < chicken.size(); i++) {
            selected[i] = true;
            sol(cnt+1, i+1);
            selected[i] = false;
        }
    }
}